components/server/src/workspace/ec2-workspace-manager.ts

```diff
@@ -0,0 +1,724 @@
+import * as crypto from 'crypto';
+import { inject, injectable } from "inversify";
+import { CommitContext, StartWorkspaceResult, User, Workspace, WorkspaceInstance, Identity } from "@gitpod/gitpod-protocol";
+import { CloneTargetMode, GitInitializer } from "@gitpod/content-service/lib";
+import { DBWithTracing, TracedWorkspaceDB, WorkspaceDB } from '@gitpod/gitpod-db/lib';
+import { log } from '@gitpod/gitpod-protocol/lib/util/logging';
+import { TraceContext } from "@gitpod/gitpod-protocol/lib/util/tracing";
+import { HostContextProvider } from "../auth/host-context-provider";
+import { Env } from "../env";
+import { SessionHandlerProvider } from '../session-handler';
+import { TokenProvider } from "../user/token-provider";
+import { MessageBusIntegration } from "./messagebus-integration";
+import { URL } from 'url';
+import * as ec2 from '@aws-sdk/client-ec2';
+import * as ssm from '@aws-sdk/client-ssm';
+import * as http from 'http';
+import * as https from 'https';
+
+interface EC2WorkspaceInstance {
+    workspaceId?: string;
+    ec2InstanceId?: string;
+};
+
+const enum EC2InstanceStatus {
+    UNDEFINED = -1,
+    PENDING = 0,
+    RUNNING = 16,
+    SHUTTING_DOWN = 32,
+    TERMINATED = 48,
+    STOPPING = 64,
+    STOPPED = 80
+};
+
+interface EC2WorkspaceDetails {
+    workspace: EC2WorkspaceInstance;
+    ec2StatusCode: EC2InstanceStatus;
+    ec2SystemStatus: boolean;
+    ec2InstanceStatus: boolean;
+};
+
+@injectable()
+export class EC2WorkspaceManager {
+    @inject(Env) protected readonly env: Env;
+    @inject(TracedWorkspaceDB) protected readonly workspaceDb: DBWithTracing<WorkspaceDB>;
+    @inject(TokenProvider) protected readonly tokenProvider: TokenProvider;
+    @inject(HostContextProvider) protected readonly hostContextProvider: HostContextProvider;
+    @inject(MessageBusIntegration) protected readonly messageBus: MessageBusIntegration;
+    @inject(SessionHandlerProvider) protected sessionHandlerProvider: SessionHandlerProvider;
+
+    // Path to store all devspaces configuration files on the instance
+    private readonly workspaceConfigDir = '~/.devspaces';
+
+    public isEC2WorkspaceInstance(instance: WorkspaceInstance): boolean {
+        const raw_regex_pattern = '^https://windows-[0-9]{1,3}-[0-9]{1,3}-[0-9]{1,3}-[0-9]{1,3}\\.ws\\.' + this.env.gitpodDomain.replace(/\./gi, "\\.");
```

Comment:
IP expression can be changed to vanity names given for a pod ?

Reply:
That is planned for the future, it requires a bit more work in the proxying. Instead of caddy doing it, I'll need to extend ws-proxy to handle the IP to slug mapping.